Feature: Agent/Department Visibility

aydreeihn opened the pull request:
This commit adds a new feature that allows admins to limit the Departments or Agents that are visible in the Agent Panel based on new permissions. The permissions can be found here:

Admin Panel | Agents | Click an Agent | Permissions Tab | Miscellaneous

<img width="749" alt="Screen Shot 2020-03-10 at 3 14 46 PM" src="https://user-images.githubusercontent.com/22985133/76355252-f1ac8380-62e1-11ea-9ab6-e33ae7ff05fa.png">

These boxes will be checked by default for new agents as well as existing agents (in the case of an upgrade).

You can check out the documentation [here](https://github.com/osTicket/docs/blob/945c7d54c800f55eb297d8b3b27337f8122954c7/Features/Visibility%20Permissions.rst).

JediKev committed:
db: Latest Indexes

This adds all indexes we've implemented/accumulated over time.

aydreeihn committed:
Filter Canned Responses Page

This commit filters Canned Responses so that Agents can only see Canned Responses for Departments they can access (Primary or Extended Access) or Canned Responses assigned to all Departments.

aydreeihn committed:
Filter Canned Response Creation

This commit makes it so that Agents can only create Canned Responses for Departments they can access or Canned Responses that will show up for all Departments.

aydreeihn committed:
Add Agent/Department Visibility Checkboxes

This commit adds the following checkboxes to Admin Panel | Agents | Choose an Agent | Permissions | Miscellaneous:
- Agent — Ability to see Agents in all Departments
- Department — Ability to see all Departments

aydreeihn committed:
Limit Departments in DepartmentField Choices

This commit checks to see if the Agent has permission to see all Departments, and if they don't, show only their Departments in forms that include a DepartmentField. This covers the following:
	- Ticket Transfers
	- Task Transfers
	- Ticket Referrals
	- Task Creation

Standard Forms:
	- Ticket Creation

aydreeihn committed:
getDepartmentNames Publiconly Option

This commit allows us to determine if only public Departments should be returned or all Departments regardless of anything else. We need this for Agents that can see all Departments and search all Tickets

aydreeihn committed:
Canned Responses for Dept Roles

This commit fixes a bug where we did not accurately check individual Roles for Departments an Agent could access to determine if they could manage that particular Canned Response.

Now, we check to see if the Agent has a Role that can manage Canned Responses at all to determine if Canned Responses show up in the Knowledge Base.

Additionally, we filter Canned Responses shown based on the Role permissions of the Agent's Departments (Primary and Extended)

aydreeihn committed:
Limit Agent Choices

This commit checks if the Agent has permission to see all Agents, and if they don't, only show Agents that are in Departments they have access to. This covers the following:
- Queue Assign Mass Process
- Ticket open Assign
- Ticket View assign
- Ticket Inline Edit Assign
- Ticket Refer
- Assigned Staff Quick Filter
- Assignee Quick Filter
- Department Manager Quick Filter

It also limits Agents shown in the Agent Directory based on the Agent permission checkbox as well as the Departments shown in the Department filter of the Agent Directory.

aydreeihn committed:
Option to Filter Criteria Choices

This commit gives us a way to determine if we should filter choices for setting field criteria. We need this to be optional because we use the same method for getting criteria choices in the Agent and Admin Panel, and for the Agent panel, we need to filter the choices based on the Agent's permission to view all Agents and Departments.

If Criteria are being set in the Admin Panel, all choices should be shown regardless of an Agents permissions.

filterVisibility needs to be applied for the following:
- AgentSelectionField
- DepartmentManagerSelectionField
- AssigneeChoiceField
- DepartmentField
- DepartmentChoiceField

aydreeihn committed:
Issue: Assignment Field Quick Filter

This commit fixes an issue where quick filters for the AssigneeField class would not work because we were returning an array of AgentsName + Team name strings which did not sanitize correctly in htmlchars.

aydreeihn committed:
Agent/Dept Visibility Defaults

This commit makes it so that the Agent and Department visibility permission is checked by default on fresh installs and newly created Agents unless otherwise specified.

aydreeihn committed:
Update Existing Staff Permissions

This commit updates existing staff members so that they will still be able to see departments and agents within the agent panel the same way they are used to seeing them. This is done during the upgrade within a task.

aydreeihn committed:
Ticket Open Canned Responses

If an Agent is opening a Ticket, they should only see Canned Responses for Departments they have access to.

aydreeihn committed:
Filter Topics Based on Dept Access

This commit makes it so that Help Topics are hidden from the Agent Panel if they are assigned a Department an Agent does not have access to and the Agent does not have the 'Ability to see all Departments' permission checked.

If a Help Topic is Public, it will be shown in the Agent Panel.

If a Parent Help Topic is Private and assigned a Department an Agent does not have access to, we hide both the Parent and Children Help Topics.

If a Help Topic is not assigned a Department at all, it is shown throughout the Agent Panel.

Help Topics are filtered for:
- Inline Ticket Edit
- Ticket Edit
- Ticket Open
- Advanced Search
- Quick Filters
- Dashboard

aydreeihn committed:
Filter FAQs Based on Topic Access

This commit makes it so that FAQs that are associated with Help Topics that are associated with Departments an Agent does not have access to are hidden. It also ensures that an Agent cannot create an FAQ associated with a Help Topic that they do not have access to based on the Help Topic's Department.

aydreeihn committed:
Clean Topics Array

This commit makes sure that we have a clean array for filtered topics so that disabled/archived help topics are correctly handled in the returned selection.

aydreeihn committed:
FAQ Topic Quick Filter

This commit makes sure that we do not show the names of any hidden Help Topics in the Quick Filter for Topics on the FAQs page in the Agent Panel.

protich commented on include/class.canned.php:
```diff
@@ -229,11 +229,25 @@ static function getIdByTitle($title) {
+            $departments[] = 0;
+            //get the responses for the agent's departments
+            $staffDepts = $thisstaff->getDepartmentNames(true);
+            foreach ($staffDepts as $k => $v)
```
Instead of foreach look I would suggest then add the zero key thereafter.
```php
$departments = array_keys($thisstaff->getDepartmentNames(true));
```

aydreeihn committed:
Code Improvements

- Use array_keys instead of a foreach loop to filter canned responses for staff
- add $options var to AssigneeField getWidget function to match parent
